### DISCUSSION TITLE
Is it possible to make a completion provider in jupyter just like in `.rs`?

### DISCUSSION THREAD

**Discussion by baiguoname** [evcxr/evcxr#326] [Ideas]
There are some needing processes I can see:
1. If I am inputting a cell in jupyter as follow:
```
let a = 1;
let b = Vec::n
``` 
It need me to put the whole cell's content into `lib.rs`, but how can I know where the cell's content start? I thought jupyter maybe provide a method. 
2. If I keep putting the cell's content  into `lib.rs` each time after the cell's content changed, does it have a performance issue?
3. If the rust_analyzer give back some completion items, how can I get it in jupyter, does jupyter provide api to do it?
I thought the imaging completion provider in evcxr is really powerful, as I used several jupyter kernels from other languages like python, julia etc, there is nothing like that.

**Comment by davidlattimore** [comment]
Evcxr should already provide completions in Jupyter. Here's a screenshot of what it looks like for me:

![image](https://github.com/evcxr/evcxr/assets/8983542/de409a70-490c-4c17-93cf-365566075cc9)

I copied the code you provided and then pressed the `tab` key.

Does that not work for you?

Are you using Jupyter notebook or Jupyter Lab? I use notebook. Jupyter Lab I don't think works as well, since they don't allow kernels to provide Javascript.

**Reply by davidlattimore** [reply]
Or maybe I misunderstood what you're asking...

**Reply by baiguoname** [reply]
![image](https://github.com/evcxr/evcxr/assets/52092777/ad4f8340-87e8-4d70-8f36-0e96ce337fe5)
Jupyterlab works fine too. But jupyterlab extention in vscode doesn't work.

**Reply by baiguoname** [reply]
It's really powerful in jupyterlab. Is it possible to use that in jupyterlab extention in vscode?

**Reply by davidlattimore** [reply]
I've never tried the vscode extension. I probably should, since I get the impression it's popular.

I remembered what it was that doesn't work in Jupyter Lab, it was the inline errors and warnings. Like this:

![image](https://github.com/evcxr/evcxr/assets/8983542/0526129b-d83d-47a9-ae0e-65e041cbd5ae)

**Reply by baiguoname** [reply]
Yes, jupyterlab can't show the error indication. It is a powerful feature, I thought using evcxr in vscode is just using  half of it.
